rfd/XXXXa-scoped-join-tokens.md

```diff
@@ -0,0 +1,250 @@
+
+ option go_package = "github.com/gravitational/teleport/api/gen/proto/go/teleport/scopes/joining/v1;joiningv1";
+
+@@ -46,8 +48,29 @@ message ScopedToken {
+
+ // ScopedTokenSpec is the specification of a scoped token.
+ message ScopedTokenSpec {
+-  // AssignedScope is the scope to which this token is assigned.
++  // The scope to which resources provisioned using this token are assigned.
+   string assigned_scope = 1;
+
+-  // TODO(fspmarshall): port relevant token features to scoped tokens.
++  // The list of roles associated with the token. They will be converted
++  // to metadata in the SSH and X509 certificates issued to the user of the
++  // token.
++  repeated string roles = 2;
++
++  // The joining method required in order to use this token.
++  // Supported joining methods for scoped tokens only include 'token'.
++  string join_method = 3;
++
++  // The list of rules applied to resources using this token.
++  // At least one rule must match the resource in order to join.
++  repeated types.TokenRule allow = 4;
++
++  // The set of labels to be automatically assigned to any provisioned resource.
++  map<string, string> assigned_labels = 5;
++
++  // The number of resources that can still be provisioned using this token.
++  int32 remaining_uses = 6;
+ }
+```
+
+### Provisioning
+
+Scoped resources will be provisioned by extending the existing `Join` RPC to
+support scoped tokens. When the join method is `token` the auth server will
+query both the existing `ProvisionTokenV2` and the new `ScopedToken` resources
+using the provided token name. If a `ScopedToken` is found then it will take
+precedence over any `ProvisionTokenV2` that might share the same name. This
+approach allows for backwards compatible provisioning where agents can be
+scoped without any knowledge of scopes themselves. The `assigned_scope` of the
+`ScopedToken` will be assigned as the scope of the resulting resource. It will
+also be attached as metadata to the resulting host certificate under the
+`AgentScope` field so that scope related access controls can be performed
+against the agent's identity rather than just restricting clients' access to
+the agent itself.
+
+### Automatic Labels
+
+Similar to assigning the provisioned resource scope from the `ScopedToken`
+used, the same token's `assigned_labels` field will be applied to the resulting
+resource directly after provisioning. Because static labels exist exclusively
+as configuration within the agent's `teleport.yaml` file, scoped token
+provisioning will use the `server_info` resource to apply
+[resource-based labels](https://goteleport.com/docs/zero-trust-access/rbac-get-started/labels/#apply-resource-based-labels)
+upon creation. This applies labels at runtime using the inventory control
+stream and does not require any new resources or RPCs. Labels will be created
+at the end of `RegisterUsingToken()` before returning the host certificate for a
+successfully provisioned resource.
```

Comment:
Are we ok with this feature being essentially just advisory and eventually consistent? i.e. an agent can still happily heartbeat a different set of labels (in addition to making its own authz decisions, which is always essentially dependent on the agent being well-behaved).

Reply:
If we wanted these to be enforced they'd have to end up in the host cert, too.